shutter/shutter.md

```diff
@@ -0,0 +1,392 @@
+- Decryption key: A key produced by the keyper set for each slot that decrypts transactions.
+- Instance ID: A unique identifier distinguishing between the various instantiations of Shutter. For Gnosis Chain it is `100001`, for Chiado it is `200001`.
+
+To illustrate some of these definitions, see the following example with eon boundaries at slot 0, 100, and 150:
+
+```
+slot:           | 0 | 1 | 2 | ... | 100 | 101 | 102 | ... | 150 | 151 |
+eon:            | 0               | 1                     | 2
+decryption key: | 0 | 1 | 2 | ... | 100 | 101 | 101 | ... | 150 | 151 | ...
+eon key:        | 0               | 1                     | 2
+keyper set:     | 0               | 1                     | 2
+```
+
+# Components
+
+This section describes the various components of the system and how they interact with each other.
+
+| Component              | Responsible Team |
+| ---------------------- | ---------------- |
+| Keypers                | Shutter          |
+| Validator              | Nethermind       |
+| Sequencer Contract     | Gnosis           |
+| Validator Registry     | Gnosis           |
+| Key Broadcast Contract | Shutter          |
+| Keyper Set Manager     | Shutter          |
+| Encrypting RPC Server  | Tbd              |
+| Decryption Key Relay   | Shutter          |
+
+## Keypers
+
+Keypers watch the Keyper Set Manager and detect at which slot they become part of a keyper set. Some time well before this slot, they start generating an eon key with an index given by the keyper set index in the Keyper Set Manager. The key generation process is out of scope of this document.
+
+Once the eon key is generated, they broadcast it via the Key Broadcast Contract.
+
+During the time the keyper set is active, the keypers continuously generate decryption keys. To this end, they listen for Gnosis Chain blocks. As soon as they see a new block, they broadcast their decryption key share for the following slot on a p2p network. In addition, keypers listen for key shares from other keypers and, when they collected a sufficient number of them, broadcast the aggregated decryption key.
+
+### Interfaces with
+
+- Key Broadcast Contract: To broadcast eon key
+- Keyper Set Manager: To check membership
+- Validator: To provide decryption keys to
+
+### Interface
+
+For each slot, active keypers provide keys to interested parties via a [libp2p gossipsub network](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/README.md). The relevant topics are `decryptionKeyShare` and `decryptionKey`. On the former, keypers broadcast decryption key shares, on the latter the corresponding aggregated decryption keys, in the form of the protocol buffers `DecryptionKeyShare` and `DecryptionKey`:
+
+```jsx
+message DecryptionKeyShare {
+    uint64 instanceID = 1;
+    uint64 eon = 2;
+    bytes epochID = 3;
+    uint64 keyperIndex = 4;
+    bytes share = 5;
+}
```

Comment:
Besides asserting message validity, gossip rules must bound the maximum message count on each topic. Are there multiple possible correct messages for the same key share? If yes another gossip rule must be:
- [IGNORE] Is the first seen message with <list if fields>